src/SFML/Window/Win32/JoystickImpl.cpp

```diff
@@ -101,6 +117,215 @@ JoystickBlacklist joystickBlacklist;
 const DWORD directInputEventBufferSize = 32;
 } // namespace
 
+// for XInput
+namespace
+{
+constexpr std::size_t xinputMaxDevices{4};
+std::vector<bool>     xInputSlots{};
+bool                  directInputNeedsInvalidation{};
+
+[[nodiscard]] sf::Joystick::Axis getAxis(std::size_t index)
+{
+    switch (index)
+    {
+        case 0:
+            return sf::Joystick::Axis::X;
+        case 1:
+            return sf::Joystick::Axis::Y;
+        case 2:
+            return sf::Joystick::Axis::Z;
+        case 3:
+            return sf::Joystick::Axis::R;
+        case 4:
+            return sf::Joystick::Axis::U;
+        case 5:
+            return sf::Joystick::Axis::V;
+        case 6:
+            return sf::Joystick::Axis::PovX;
+        case 7:
+            return sf::Joystick::Axis::PovY;
+        default:
+            throw std::out_of_range("index is out of range.");
+    }
+}
+
+struct XInputCleanupData
```

Comment:
If you add a destructor in this struct, and have that do what `safeCleanup` is doing,
it will be done automatically on all paths leaving the function where it is used.

Reply:
Great suggestion, taken and implemented. (Can you check that I did it right?)